backend/entities/oauth/base/service.go

```diff
@@ -0,0 +1,184 @@
+	
+	return &authorizeURL, nil
+}
+
+func (e *OAuthService) Token(userID uuid.UUID, tokenBody models.OAuthTokenRequestBody, resource models.OAuthResource) error {
+	// Parse the resource client:
+	client, err := e.parseResourceClient(resource)
+	if err != nil {
+		return err
+	}
+
+	// Retrieve the CSRF Token from the database:
+	oauthToken, err := GetOAuthToken(e.ServiceParams.DB, userID, resource)
+	if err != nil {
+		return err
+	}
+
+	// Validate the CSRF Token:
+	if oauthToken.CSRFToken != tokenBody.State || oauthToken.CSRFToken == "" {
+		return errors.New("invalid CSRF token")
+	}
+
+	// Exchange the code for an access token:
+	token, err := e.tokenExchange(client, tokenBody.Code)
+	if err != nil {
+		return err
+	}
+
+	// Save the refresh token to the database:
+	if err := SaveOAuthTokens(e.ServiceParams.DB, userID, *token, resource); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *OAuthService) Revoke(userID uuid.UUID, resource models.OAuthResource) error {
+	// Parse the resource client:
+	client, err := e.parseResourceClient(resource)
+	if err != nil {
+		return err
+	}
+
+	// Retrieve access token:
+	oauthToken, err := GetOAuthToken(e.ServiceParams.DB, userID, resource)
+	if err != nil {
+		return err
+	}
+	if oauthToken.AccessToken == "" {
+		return errors.New("user does not have an access token")
+	}
+
+	// Revoke the token on resource server:
+	if err := client.ResourceClient.Revoke(oauthToken.AccessToken); err != nil {
+		return err
+	}
+
+	// Remove refresh token from the database:
+	if err := DeleteOAuthToken(e.ServiceParams.DB, userID, resource); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *OAuthService) parseResourceClient(resourceType models.OAuthResource) (*oauth.OAuthClient, error) {
+	var resourceClient oauth.OAuthResourceClientInterface
+	switch resourceType {
+		case "google":
+			resourceClient = &e.ServiceParams.Integrations.OAuth.Google
+		case "outlook":
+			resourceClient = &e.ServiceParams.Integrations.OAuth.Outlook
+		default:
+			return nil, errors.New("invalid resource type")
+	}
+
+	return &oauth.OAuthClient{
+		ResourceClient: resourceClient,
+	}, nil
+}
+
+func (e *OAuthService) tokenExchange(client *oauth.OAuthClient, code string) (*models.OAuthToken, error) {
+	// Retrieve the client config
+	config := client.ResourceClient.GetConfig()
+	tokenUrl := fmt.Sprintf("%s/token", config.TokenURL)
+
+	fmt.Println(tokenUrl)
+
+	// Create the form data:
+	formData := url.Values{
+		"client_id": {config.ClientID.Expose()},
+		"client_secret": {config.ClientSecret.Expose()},
+		"code": {code},
+		"redirect_uri": {config.RedirectURI},
+		"grant_type": {"authorization_code"},
+	}
+
+	encodedData := formData.Encode();
+
+	// Create a new outgoing request:
+	req, err := http.NewRequest("POST", tokenUrl, strings.NewReader(encodedData))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
```

Comment:
i can help clean this up with some new helpers in [backend/utilities/http.go](https://github.com/GenerateNU/sac/blob/main/backend/utilities/http.go)